### PR TITLE
Porting l10n_ch_scan_bvr to Odoo V8 using the new API system

### PR DESCRIPTION
**Pull request by open-net-sarl**

### CODE REVIEW

**Comment by vrenaville** [comment]
Hello,
Thanks for your work
But I can't see the V7 behaviour of the default product on partner
to have a default invoice line with this product.
(https://github.com/OCA/l10n-switzerland/blob/7.0/l10n_ch_scan_bvr/partner.py)
Could you please add it ?

Kind regards,

Vincent

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+#         return self._create_direct_invoice(cr, uid, datas, context=context)
+
+    # ------------------------- Tools
+
+    def _check_number(self, part_validation):
+        nTab = [0, 9, 4, 6, 8, 2, 7, 1, 3, 5]
+        resultnumber = 0
+        for number in part_validation:
+            resultnumber = nTab[(resultnumber + int(number) - 0) % 10]
+        return (10 - resultnumber) % 10
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Please use https://github.com/odoo/odoo/blob/8.0/openerp/tools/misc.py#L579 instead : `from openerp.tools import mod10r`

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+            }
+            break
+
+        return {'value': vals}
+
+#     @api.one
+#     def do_create_invoice(self):
+#         cr, uid, context = self.env.args
+#         datas = self.read(['account_id', 'bvr_string', 'journal_id'])
+#         return self._create_direct_invoice(cr, uid, datas, context=context)
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Please remove the comment.

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+
+#         view_name = {
+#             'in_invoice': 'account.invoice.supplier.form',
+#             'out_invoice': 'account.invoice.form'
+#         }[inv_type]
+#         cr.execute(
+#             'select id,name from ir_ui_view where model=%s and name=%s',
+#             ('account.invoice', view_name)
+#         )
+#         view_res = cr.fetchone()
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Remove commented code.

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+        }[inv_type]
+
+        ctx = "{'type':'%s', 'journal_type': '%s'}" % (inv_type, journal_type)
+        return {
+            'domain': "[('id', 'in', ["+','.join(map(str, invoices))+"])]",
+            'name': 'Invoices',
+            'view_type': 'form',
+            'view_mode': 'tree, form',
+            'res_model': 'account.invoice',
+            # 'view_id': view_res,
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: More comments to remove.

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+from openerp.tools.misc import UpdateableStr
+
+import logging
+_logger = logging.getLogger(__name__)
+
+FORM = UpdateableStr()
+
+
+class onsp_wizard_scan_bvr(models.TransientModel):
+    _name = 'onsp.wiz.scan_bvr'
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Why was the module renamed from `scan.bvr` ?

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+
+from openerp import models, fields, api, _
+from openerp.exceptions import except_orm
+import time
+from openerp.tools.misc import UpdateableStr
+
+import logging
+_logger = logging.getLogger(__name__)
+
+FORM = UpdateableStr()
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Why is this used for?

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+            'view_type': 'form',
+            'view_mode': 'tree, form',
+            'res_model': 'account.invoice',
+            # 'view_id': view_res,
+            'context': ctx,
+            'type': 'ir.actions.act_window',
+            'res_id': invoices
+        }
+
+    def _create_bvr_account(self, account_unformated):
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py:318]: s/formated/formatted (also for unformatted)

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+            bvr_struct['currency'] = 'CHF'
+        #
+        elif bvr_type == '01':
+            # This BVr is the type of BVR in CHF
+            # WE will call the function and Call
+            bvr_struct = self._construct_bvr_in_chf(bvr_string)
+            # We will test if the BVR have an Adherent Number if not
+            # we will make the search of the account base on
+            # his name non base on the BVR adherent number
+#             if (bvr_struct['bvrnumber'] == '000000'):
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Remove commented code from the 7.0 version.

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+            cr, uid, data['journal_id'][0], context=context)
+        inv_type = valid_types.get(journal.type, False)
+        if not inv_type:
+            raise except_orm(
+                _('Journal Type Error'),
+                _('Invalid journal selected: must be for purchases or sales!'))
+
+        return 'create'
+
+onsp_wizard_scan_bvr()
```
Comment (mdietrichc2c) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Not needed in 8.0 (or 7.0)

**Comment by yvaucher** [comment]
Hello @open-net-sarl thanks for you proposal.
However it cannot be accepted in this current state :-1: 

I will gladly review it once you have fix those few things (I will set it as WIP in the mean time):

**1) This module needs to be based on top of the version 7.0, moved from `__unported__`
https://github.com/OCA/l10n-switzerland/tree/7.0/l10n_ch_scan_bvr .**

Please do this it in at least 2 commits:
- one to move all files from `__unported__`
  
  ```
  git mv __unported__/l10n_ch_scan_bvr/ .
  git commit -m '<some message telling you moved it from unported>'
  ```
- and other to do your changes
  
  It is preferable to have those 2 operations separated to keep a clean history

**2) You have added your copyright in a strange manner removing the former ones.**

Please restore original authors: @nbessi and @vrenaville and remove your "Created by" as it is a copy.

**3) module icon and some class and method names have company specific names.**

As OCA is managed by a community, we want to limit the mention to companies to authors. This in order to encourage contributions.

**4) there are some commented code to clean up**

Keep your code base clean git is there to remind of old code we don't use anymore.

**5) licenses headers must be AGPL not GPL**

This might come from the old version of the module in 6.1 however we want to continue the work done in 7.0, and at that time license was changed in agreement with the authors from GPL to AGPL. The term "Affero" must be present in header.

File: l10n_ch_scan_bvr/wizard/scan_bvr_view.xml [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                            <field name="account_id" domain="[('partner_id', '=', partner_id),('state', 'like', 'bvr')]"/>
+                            <field name="amount" widget="monetary" options="{'currency_field': 'currency_id'}" readonly="1"/>
+                        </group>
+                    </group>
+                    <separator string="" colspan="4"/>
+    				<group col="14" colspan="4">
+                        <label string="" colspan="10"/>
+    					<button special="cancel" string="Cancel" icon="gtk-cancel" colspan="2"/>
+    					<button name="do_create_invoice" string="Validate" type="object" icon="gtk-ok" colspan="2"/>
+    				</group>
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr_view.xml]: Please replace those tabs by spaces

**Comment by open-net-sarl** [comment]
@yvaucher thanks for the reply. However, and as I explain a few days ago: I am still learning git, so sorry if the question seems obvious but before executing "git mv .... etc", must I remove everything? or do a checkout?

**Comment by yvaucher** [comment]
@open-net-sarl What you can do is clone a fresh 8.0 branch, move module from `__unported__` and commit then start to add your changes.

3 options for that:
- git rebase the branch of this PR on top of the fresh branch with moved module from `__unported__` and resolve all conflicts
- OR copy/past your module to overwrite the 7.0 in the fresh branch and use `git diff` to know which change to keep or not, you can even use `git add -p` to select the changes you want and the one you don't want.
- OR from 7.0 version of module patch it with your changes.

Second options might be the best one `git add -p` should help a lot.

After having done those changes you can do a `git push -f` to update this PR

**Comment by yvaucher** [comment]
In fact there is already a portage made there:
https://github.com/OCA/l10n-switzerland/pull/98/files

**Comment by open-net-sarl** [comment]
@yvaucher thanks for the last link but: it hasn't changed since a few months. So what should I do now?
As you explained sooner: start a new branch?

**Comment by yvaucher** [comment]
@open-net-sarl 

@hurrinico [said](https://github.com/OCA/l10n-switzerland/pull/98#issuecomment-137753707) he would work on it ASAP, please ask him directly. In case OpenNet have more time than him I would advise to take over his PR #98  to add your changes on top of it.

**Comment by open-net-sarl** [comment]
Hello,

We already started to work on it this morning, from a fresh pull of the unported

We will check the https://github.com/OCA/l10n-switzerland/pull/98 to include whats needed

And hopefully, pull the whole thing later today, with the uses of the default product.

Hope it ok with everybody and the GIT/OCA process.

**Comment by open-net-sarl** [comment]
Hello,

As I said, we try to all features in this new pull

Thanks to
https://github.com/yvaucher
https://github.com/hurrinico
for upcoming review

File: l10n_ch_scan_bvr/README.rst [hunk 11]
```diff
@@ -0,0 +1,43 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/README.rst]: Please have a look at template it has evolved a bit:

https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

**Comment by yvaucher** [comment]
@open-net-sarl As when porting a module the first reflex should be to have a look on github to see if someone is already doing it, I would prefer that you base your work on branch of #98 as a superseeded work taking commits done by @hurrinico 

Plus I'm also concerned about some author change of the module like README.rst writing in log that you ported it to v7.0 while v7.0 is existing and was ported by @vrenaville . It might be right you did a private version of it in 7.0, however it exists already in 7.0 and this one wasn't ported by you.

I see that you also change things like website, in `__openerp__.py` , I understand that you change it to deliver it like that to you customer, as long as you keep original copyrights it's ok, but I don't see it good here on community version.

So let try to clean this contribution:

What I would like to see in this PR is list of commits that way (use `git push -f` to fix):
- A commit moving files from `__unported__` to root for the ported module.
- A commit with changes made by @hurrinico in #98 
- 1 or multiple commits with your changes

**Comment by open-net-sarl** [comment]
I think we have to keep this simple

The Pull #98 was left untouched since february in a state that failed the Travis test

We asked on the OCA mailing list a week ago how to deal with a new Pull for the scan BVR
We got no answer but we read the discussion on OCA about what to do with old Pull, and the idea was to closed them up

So we didn't start from Pull #98, left in a state that failed Travis test

We made a first Pull and we received some worthwhile advices to based on the 'unported' and to add the default product.

So we did, as well as some minor typo in the __openerp file

We want this feature ( scan BVR ) to be available to all swiss users, we want to have it ported from A (unported) to B (ported) state.
I don't understand why the process should be complicated by adding changes from #98, ( going to C first ) that will anyway been overidden by the changes of #163

We don't want to take over the work of anybody else, we want to have a working, useable version of this module in the l10n-switzerland. And I think it is what this Pull is about.

**Comment by yvaucher** [comment]
The reason behind having one commit for file move only is because of how git is made, the risk is that it might not be able to follow a moving file in history if a big part of it has been modified which is the case with API 8 migration. All other commits before that "moved from `__unported__` " commit don't make any sense and just make the history unclean. Plus having one commit only with `__unported__` make easier the review process as the reviewer can see a diff of other commits knowing what was added in the migration process.

The reason why making the work on top of @hurrinico is because it is closest to the 7.0 version than yours. So what was fixed in 7.0 of the community version is more likely in his version.  As you said it has been there for a while. If it hasn't move since a long time it is certainly that it was already a working version. Plus, if you were aware of this existing PR you could have let a simple and little message in it to get some answer about the state of it. Or saying something like "hey we saw your work and are willing to finish the port we will superseed it is that ok for you?". Then having your changes based on @hurrinico work would tell us what is the difference between the two versions.

I definitively want it to land, but there is no reason to rush and I can't accept it as long as it doesn't respect the OCA guidelines, no one seems to be in the hurry as #98 didn't get a lot of attention.

If you want to keep it simple having a cleaner history would help. Here the history won't be simple.

Furthermore to start reviewing I need to know what changed and it is far from simple without the little move from `__unported__` as a first commit. May I help you with that?

**Comment by open-net-sarl** [comment]
KISS Principle

" A commit with changes made by @hurrinico in #98 "

You asking us to pass through an intermediate step using a 7 months left over pull, this does not make sense to us, so we closed the Pull. It's cleaner.

Anyway, as you said " no one seems to be in the hurry "

In V9, I heard that there will no unported, no obsolete Pull, so we might give it another shot.
For the V8, we leave to Pull #98 

Sorry to the community for not having a little usefull tool.

**Comment by yvaucher** [comment]
Here we go, 20 minutes and its done, see here the result:
https://github.com/yvaucher/l10n-switzerland/tree/8.0-port-l10n_ch_scan_bvr

Super simple and super fast :smile: 

Here are the commands I did:

```
git clone git@github.com:oca/l10n-switzerland
git checkout origin/8.0 -b 8.0-port-l10n_ch_scan_bvr
git mv __unported__/l10n_ch_scan_bvr/ .
git commit -m 'Moved l10n_ch_scan_bvr out of __unported__
git checkout origin/8.0 -b hurrinico-adding_8.0_scan_bvr
git pull https://github.com/hurrinico/l10n-switzerland.git adding_8.0_scan_bvr
git rebase -i origin/8.0
git log -n 1
# get the commit hash
git cherry-pick hash
cd ..
git clone git@github.com:open-net-sarl/l10n-switzerland on-l10n
cd on-l10n
git checkout l10n_ch_scan_bvr-ported-to-v8
cp -fr l10n_ch_scan_bvr ../l10n-switzerland/
cd ../l10n-switzerland
git status
# it shows:
# On branch 8.0-port-l10n_ch_scan_bvr
# Your branch is ahead of 'origin/8.0' by 2 commits.
#   (use "git push" to publish your local commits)
# Changes not staged for commit:
#   (use "git add <file>..." to update what will be committed)
#   (use "git checkout -- <file>..." to discard changes in working directory)
# 
#   modified:   l10n_ch_scan_bvr/README.rst
#   modified:   l10n_ch_scan_bvr/__init__.py
#   modified:   l10n_ch_scan_bvr/__openerp__.py
#   modified:   l10n_ch_scan_bvr/i18n/fr.po
#   modified:   l10n_ch_scan_bvr/wizard/__init__.py
#   modified:   l10n_ch_scan_bvr/wizard/scan_bvr.py
#   modified:   l10n_ch_scan_bvr/wizard/scan_bvr_view.xml
# 
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
# 
#   l10n_ch_scan_bvr/models/
#   l10n_ch_scan_bvr/static/
#   l10n_ch_scan_bvr/views/

git add l10n_ch_scan_bvr
git commit --author="Open Net Ltd <info@open-net.ch>" -m 'Porting l10n_ch_scan_bvr to Odoo V8 using the new API system'
```

And as you can see in commits it's now clean:
https://github.com/yvaucher/l10n-switzerland/commits/8.0-port-l10n_ch_scan_bvr

We for example see that we need to work a bit on the translation file full of empty strings.

If you want to keep the good work you might want to take my branch cloning it or using git remote to replace this one, then I will be able to start the true review. You can also start a new PR if you prefer.

**Pull request closed without merging**